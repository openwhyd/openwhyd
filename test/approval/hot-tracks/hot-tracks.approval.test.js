// Run with: $ npm run test:approval:hot-tracks
//   ... or: $ npm run test:approval:hot-tracks -- --updateSnapshot

const waitOn = require('wait-on');
const {
  httpClient,
  ObjectId,
  connectToMongoDB,
  OpenwhydTestEnv,
  indentJSON,
  getCleanedPageBody,
} = require('../../approval-tests-helpers');

const {
  START_WITH_ENV_FILE,
  PORT, // Note: if PORT is not provided, approval-tests-helpers will start Openwhyd's server programmatically, using START_WITH_ENV_FILE
  DONT_KILL,
} = process.env;

const openwhyd = new OpenwhydTestEnv({
  startWithEnv: START_WITH_ENV_FILE,
  port: PORT,
});

const MONGODB_URL =
  process.env.MONGODB_URL || 'mongodb://localhost:27117/openwhyd_test';

const users = [
  {
    _id: ObjectId('61e19a3f078b4c9934e72ce1'),
    name: 'user 0',
    email: 'users[0]@test.com',
    pwd: '21232f297a57a5a743894a0e4a801fc3',
  },
  {
    _id: ObjectId('61e19a3f078b4c9934e72ce2'),
    name: 'user 1',
    email: 'users[1]@test.com',
    pwd: '21232f297a57a5a743894a0e4a801fc3',
  },
];

const tracks = [
  {
    _id: ObjectId('61e19a3f078b4c9934e72ce6'),
    eId: '/yt/track_A',
    score: 0,
  },
  {
    _id: ObjectId('61e19a3f078b4c9934e72ce7'),
    eId: '/yt/track_B',
    score: 0,
  },
];

const loginUsers = (server, users) =>
  Promise.all(
    users.map(({ email, pwd }) =>
      httpClient.get({
        url: `${server.getURL()}/login?action=login&ajax=1&email=${email}&md5=${pwd}`,
      }),
    ),
  );

// eslint-disable-next-line @typescript-eslint/no-unused-vars
async function postTrack(server, userSession, { _id, ...trackData }) {
  const { body } = await httpClient.post({
    url: `${server.getURL()}/api/post`,
    body: { action: 'insert', ...trackData },
    cookies: userSession.cookies,
  });
  return JSON.parse(body);
}

describe('Hot Tracks (approval tests - to be replaced later by unit tests)', () => {
  /** @type import('mongodb').MongoClient */
  let mongoClient;
  /** @type import('mongodb').Db */
  let db;

  beforeAll(async () => {
    await openwhyd.setup();
    if (PORT) {
      await waitOn({ resources: [`http://localhost:${PORT}`], timeout: 1000 });
    }
    // if this test times out, make sure to start MongoDB first: $ docker-compose up -d mongo
    mongoClient = await connectToMongoDB(MONGODB_URL);
    db = await mongoClient.db();
  });

  afterAll(async () => {
    if (mongoClient) await mongoClient.close();
    if (!DONT_KILL) await openwhyd.release();
  });

  beforeEach(async () => {
    await openwhyd.reset(); // prevent side effects between tests by resetting db state
    await db.collection('user').deleteMany({}); // clear users
    await db.collection('post').deleteMany({}); // clear posts
    await db.collection('track').deleteMany({}); // clear tracks
  });

  it('renders a limited number of ranked tracks', async () => {
    await db.collection('track').insertMany([
      {
        _id: ObjectId('61e19a3f078b4c9934e72ce4'),
        name: 'a regular track',
        score: 1,
      },
      {
        _id: ObjectId('61e19a3f078b4c9934e72ce5'),
        name: 'a popular track',
        score: 2,
      },
    ]);
    await openwhyd.refreshCache();
    const json = await httpClient.get({
      url: `${openwhyd.getURL()}/hot?limit=1&format=json`,
    });
    expect(indentJSON(json.body)).toMatchSnapshot();
    const html = await httpClient.get({
      url: `${openwhyd.getURL()}/hot?limit=1&`,
    });
    expect(getCleanedPageBody(html.body)).toMatchSnapshot();
  });

  it('renders ranked tracks, starting at a given index', async () => {
    await db.collection('track').insertMany([
      {
        _id: ObjectId('61e19a3f078b4c9934e72ce4'),
        name: 'a regular track',
        score: 1,
      },
      {
        _id: ObjectId('61e19a3f078b4c9934e72ce5'),
        name: 'a popular track',
        score: 2,
      },
    ]);
    await openwhyd.refreshCache();
    const json = await httpClient.get({
      url: `${openwhyd.getURL()}/hot?skip=1&format=json`,
    });
    expect(indentJSON(json.body)).toMatchSnapshot();
    const html = await httpClient.get({
      url: `${openwhyd.getURL()}/hot?skip=1&`,
    });
    expect(getCleanedPageBody(html.body)).toMatchSnapshot();
  });

  it('renders posts enriched with track metadata', async () => {
    const posts = [
      {
        _id: ObjectId('61e19a3f078b4c9934e72ce4'),
        eId: tracks[0].eId,
        pl: { name: 'soundtrack of my life', id: 0 }, // metadata from the post that will be included in the list of hot tracks
      },
      {
        _id: ObjectId('61e19a3f078b4c9934e72ce5'),
        eId: tracks[1].eId,
        text: 'my favorite track ever!', // metadata from the post that will be included in the list of hot tracks
      },
    ];
    await db
      .collection('track')
      .insertMany(
        tracks.map((_, i) => ({ ...tracks[i], pId: posts[i]._id, score: i })),
      );
    await db.collection('post').insertMany(posts);
    await openwhyd.refreshCache();
    const json = await httpClient.get({
      url: `${openwhyd.getURL()}/hot?format=json`,
    });
    expect(indentJSON(json.body)).toMatchSnapshot();
    const html = await httpClient.get({ url: `${openwhyd.getURL()}/hot` });
    expect(getCleanedPageBody(html.body)).toMatchSnapshot();
  });

  it("updates the score of a track when it's liked", async () => {
    await db.collection('user').insertMany(users);
    await db.collection('track').insertMany(tracks);
    await openwhyd.refreshCache();
    const userSession = await loginUsers(openwhyd, users);
    // user 0 posts track A
    const { _id } = await postTrack(openwhyd, userSession[0], tracks[0]);
    const cleanJSON = (body) => body.replaceAll(_id, '__posted_track_id__');
    // user 1 likes track A
    await httpClient.post({
      url: `${openwhyd.getURL()}/api/post/${_id}`,
      body: { action: 'toggleLovePost' },
      cookies: userSession[1].cookies,
    });
    await new Promise((resolve) => setTimeout(resolve, 1000)); // give time for track model to take the like into account
    const json = await httpClient.get({
      url: `${openwhyd.getURL()}/hot?format=json`,
    });
    expect(cleanJSON(indentJSON(json.body))).toMatchSnapshot();
    // Note: the requests above mutate data => we snapshot the state of the "tracks" table.
    const tracksCollection = await db.collection('track').find({}).toArray();
    expect(cleanJSON(indentJSON(tracksCollection))).toMatchSnapshot();
  });

  it("updates the score of a track when it's reposted", async () => {
    await db.collection('user').insertMany(users);
    await db.collection('track').insertMany(tracks);
    await openwhyd.refreshCache();
    const userSession = await loginUsers(openwhyd, users);
    // user 0 posts track A
    const { _id } = await postTrack(openwhyd, userSession[0], tracks[0]);
    const cleanJSON = (body) => body.replaceAll(_id, '__posted_track_id__');
    // user 1 reposts track A
    await httpClient.post({
      url: `${openwhyd.getURL()}/api/post`,
      body: { action: 'insert', pId: _id },
      cookies: userSession[1].cookies,
    });
    await new Promise((resolve) => setTimeout(resolve, 1000)); // give time for track model to take the repost into account
    const json = await httpClient.get({
      url: `${openwhyd.getURL()}/hot?format=json`,
    });
    expect(cleanJSON(indentJSON(json.body))).toMatchSnapshot();
    // Note: the requests above mutate data => we snapshot the state of the "tracks" table.
    const tracksCollection = await db.collection('track').find({}).toArray();
    expect(cleanJSON(indentJSON(tracksCollection))).toMatchSnapshot();
  });
});
