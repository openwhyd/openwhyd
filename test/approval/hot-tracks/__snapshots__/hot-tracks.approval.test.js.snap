// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hot Tracks renders posts enriched with track metadata 1`] = `
"{
  hasMore: false,
  tracks: [
    {
      _id: '61e19a3f078b4c9934e72ce5',
      eId: '/yt/track_B',
      img: null,
      name: 'a popular track',
      nbL: 0,
      nbR: 3,
      pId: '61e19a3f078b4c9934e72ce5',
      pl: null,
      score: 3,
      trackUrl: '//youtube.com/watch?v=track_B',
      uId: null,
      uNm: null
    },
    {
      _id: '61e19a3f078b4c9934e72ce4',
      eId: '/yt/track_A',
      img: null,
      name: 'a regular track',
      nbL: 0,
      nbR: 2,
      pId: '61e19a3f078b4c9934e72ce4',
      pl: {
        id: 0,
        name: 'soundtrack of my life'
      },
      score: 2,
      trackUrl: '//youtube.com/watch?v=track_A',
      uId: null,
      uNm: null
    }
  ]
}"
`;

exports[`Hot Tracks renders posts enriched with track metadata 2`] = `
"<!DOCTYPE html>
<html lang="en">
  <head prefix="og: http://ogp.me/ns# fb: http://ogp.me/ns/fb# whydapp: http://ogp.me/ns/fb/whydapp#">
    <meta charset="utf-8" />
<meta name="google-site-verification" content="mmqzgEU1bjTfJ__nW6zioi7O9vuur1SyYfW44DH6ozg" />
<link rel="image_src" href="http://localhost:8080/images/logo-black-square-smaller.png"/>
<meta name="description" content="Discover and collect music gems from Youtube, Soundcloud, Deezer and more" />
<meta name="keywords" content="discover, music, curation, streaming, tracks, youtube, soundcloud, bandcamp, playlists, play, free" />
<meta name="twitter:card" content="summary" />
<meta name="twitter:site" content="@open_whyd" />
<meta property="og:image" content="http://localhost:8080/images/logo-black-square-smaller.png" />
<meta property="og:description" content="Discover and collect music gems from Youtube, Soundcloud, Deezer and more" />
<meta property="og:type" content="website" />
    <link href="/favicon.ico" rel="shortcut icon" type="image/x-icon" />
    <link href="/favicon.png" rel="icon" type="image/png" />
    <link href='//fonts.googleapis.com/css?family=Varela+Round' rel='stylesheet' type='text/css'>
    <link href='//fonts.googleapis.com/css?family=Varela' rel='stylesheet' type='text/css'>
    <link rel="search" type="application/opensearchdescription+xml" title="Whyd" href="http://localhost:8080/html/opensearch.xml">
    <link rel="chrome-webstore-item" href="https://chrome.google.com/webstore/detail/foohaghobcolamikniehcnnijdjehfjk">

<script>
  window.user = { id: undefined, name: '', handle: '', pref: {}, lastFm: undefined };
  window.playTrack = window.playTrack || function(){};
</script>
<script src="/js/whydtr.js"></script>
<link rel="stylesheet" type="text/css" href="/css/cookieconsent2-3.0.3.min.css" />
<script src="/js/cookieconsent2-3.0.3.min.js"></script>
<script>
  /* generated from https://cookieconsent.insites.com/download/ */
  window.addEventListener("load", function(){
    window.cookieconsent.initialise({
      palette: { popup: { background: "#000" }, button: { background: "#f1d600" } }
    })
  });
</script>
    <title>Openwhyd â€“ Discover and collect the best music tracks from the web</title>
    <link href="/css/common.css" rel="stylesheet" type="text/css" />
    <link href="/css/browse.css" rel="stylesheet" type="text/css" />
    <link href="/css/tipsy.css" rel="stylesheet" type="text/css" />
    <link href="/css/userProfileV2.css" rel="stylesheet" type="text/css" />
    <link href="/css/userPlaylistV2.css" rel="stylesheet" type="text/css" />
    <link href="/css/dlgEditProfileCover.css" rel="stylesheet" type="text/css" />
    <script src="/js/jquery-1.10.2.min.js"></script>
    <script src="/js/jquery-migrate-1.2.1.js"></script>
    <script src="/js/soundmanager2-nodebug-jsmin.js"></script>
    <script>soundManager.setup({url: "/swf/", flashVersion: 9, onready: function() {soundManager.isReady=true;}});</script>
  </head>
  <body class="pgHotTracks visitor">
   <div id="fb-root"></div>
<!--[if lt IE 8]>
<div class="topWarning">Warning: your web browser is not supported by Openwhyd. Please upgrade to a modern browser.</div>
<![endif]-->
<div id="header"><div class="container">
  <div id="headCenter">
   <a id="logo" title="Openwhyd" target="_top" class="homeLink" href="/">
   </a>

  </div>
  <div id="navbar">
   <a target="_top" id="tabStream" href="/">Stream</a>
   <a target="_top" id="tabHot" href="/hot">Hot Tracks</a>
  </div>
  <div id="searchBar">
   <div class="searchForm" id="searchForm">
    <input name="q" class="q search-input" id="q" type="text" value="" placeholder="Paste a YouTube / Soundcloud / Bandcamp / Deezer URL" autocomplete="off" />
   </div>
   <div class="searchResults" id="searchResults"></div>
  </div>
  <div id="homeHeader">
   <h1>Openwhyd: The community of music lovers</h1>
	  <p>Discover and collect music gems from Youtube, Soundcloud, Deezer and more</p>
  </div>
  <div id="logBox">
   <a id="signin" href="/login">Login</a>
   <a id="signup" onclick="login();">Sign up</a>
  </div>
</div></div>
<div id="contentPane">
  <div id="mainPanel">
<div class="container">
  <div id="filterSelector">
    <ul>
      <li class="selected">
        <a href="#">All</a>
      </li>
    </ul>
    <div id="sideFooter">
      <ul>
        <li>
          <a target="_blank" href="https://github.com/openwhyd">GitHub</a>
        </li>
        <li>
          <a target="_blank" href="http://www.facebook.com/openwhyd"
            >Facebook</a
          >
        </li>
        <li>
          <a target="_blank" href="http://twitter.com/open_whyd">Twitter</a>
        </li>
        <li>
          <a
            target="_blank"
            href="https://github.com/openwhyd/openwhyd/blob/main/docs/FAQ.md"
            >FAQ</a
          >
        </li>
      </ul>
      <p>Openwhyd is maintained collaboratively,<br />under the MIT License</p>
    </div>
  </div>
  <div class="sub-container">
    <h2 class="head-name-tab">
      Hot Tracks
      <span class="genreSelector">/ All</span>
    </h2>
    <div class="posts">
       <div
  class="post "
  data-pid="61e19a3f078b4c9934e72ce5"
  data-score="3"
  data-order=""
  data-initialpid=""
  data-loved=""
  data-time="1642175039000"
>
  <div class="playBar"></div>

  <a
    class="thumb"
    href="//youtube.com/watch?v=track_B"
    target="_blank"
    data-eid="/yt/track_B"
    onclick="return playTrack(this);"
    style="background-image:url('/images/cover-track.png');"
  >
    <img src="/images/cover-track.png" />
    <div class="play"></div>
  </a>

  <h2>
    <a
      href="/c/61e19a3f078b4c9934e72ce5"
      class="no-ajaxy"
      onclick="toggleComments('61e19a3f078b4c9934e72ce5');return false;"
      >a popular track</a
    >
  </h2>

  <p class="author">
    <span style="background-image:url('/img/u/');"></span>
    <a href="/u/"></a>  
      </p>

  <span class="ago" datetime="1642175039000">
    <a href="/c/61e19a3f078b4c9934e72ce5">(age)</a>
  </span>

  <div class="stats">
    <span
      class="nbReposts"
      style=""
      onclick="javascript:showReposts('61e19a3f078b4c9934e72ce5')"
    >
      <span>3</span>
    </span>
    <span
      class="nbLoves"
      style="display:none;"
      onclick="javascript:showPostLovers('61e19a3f078b4c9934e72ce5')"
    >
      <span>0</span>
    </span>
    <span
      class="nbComments"
      style="display:none;"
      onclick="toggleComments('61e19a3f078b4c9934e72ce5');"
    >
      <span>0</span>
    </span>
  </div>

  <div class="btns">
    <a class="btnRepost" href="javascript:publishPost('61e19a3f078b4c9934e72ce5');"> Add to </a>
    <span>&middot;</span>
    <a
      class="btnLike "
      href="javascript:toggleLovePost('61e19a3f078b4c9934e72ce5');"
    >
       Like
    </a>
    <span>&middot;</span>
    <a class="btnComment" href="javascript:toggleComments('61e19a3f078b4c9934e72ce5');">
      Comment
    </a>
    <span>&middot;</span>
    <a class="btnShare" href="javascript:sharePost('61e19a3f078b4c9934e72ce5');">
      <span> Share </span>
    </a>
    <span class="postEdit">
      <span>&middot;</span>
      <a onclick="modalPostEditBox('61e19a3f078b4c9934e72ce5')"> Edit </a>
    </span>
  </div>

  <div class="ext hidden">
    <p>
      15:43 - 14 Jan 2022  &middot; Via
      <a class="via youtube" href="//youtube.com/watch?v=track_B">YouTube</a>
           </p>

    <form class="comments">
    </form>
  </div>
</div>
<div
  class="post "
  data-pid="61e19a3f078b4c9934e72ce4"
  data-score="2"
  data-order=""
  data-initialpid=""
  data-loved=""
  data-time="1642175039000"
>
  <div class="playBar"></div>

  <a
    class="thumb"
    href="//youtube.com/watch?v=track_A"
    target="_blank"
    data-eid="/yt/track_A"
    onclick="return playTrack(this);"
    style="background-image:url('/images/cover-track.png');"
  >
    <img src="/images/cover-track.png" />
    <div class="play"></div>
  </a>

  <h2>
    <a
      href="/c/61e19a3f078b4c9934e72ce4"
      class="no-ajaxy"
      onclick="toggleComments('61e19a3f078b4c9934e72ce4');return false;"
      >a regular track</a
    >
  </h2>

  <p class="author">
    <span style="background-image:url('/img/u/');"></span>
    <a href="/u/"></a>  
      </p>

  <span class="ago" datetime="1642175039000">
    <a href="/c/61e19a3f078b4c9934e72ce4">(age)</a>
  </span>

  <div class="stats">
    <span
      class="nbReposts"
      style=""
      onclick="javascript:showReposts('61e19a3f078b4c9934e72ce4')"
    >
      <span>2</span>
    </span>
    <span
      class="nbLoves"
      style="display:none;"
      onclick="javascript:showPostLovers('61e19a3f078b4c9934e72ce4')"
    >
      <span>0</span>
    </span>
    <span
      class="nbComments"
      style="display:none;"
      onclick="toggleComments('61e19a3f078b4c9934e72ce4');"
    >
      <span>0</span>
    </span>
  </div>

  <div class="btns">
    <a class="btnRepost" href="javascript:publishPost('61e19a3f078b4c9934e72ce4');"> Add to </a>
    <span>&middot;</span>
    <a
      class="btnLike "
      href="javascript:toggleLovePost('61e19a3f078b4c9934e72ce4');"
    >
       Like
    </a>
    <span>&middot;</span>
    <a class="btnComment" href="javascript:toggleComments('61e19a3f078b4c9934e72ce4');">
      Comment
    </a>
    <span>&middot;</span>
    <a class="btnShare" href="javascript:sharePost('61e19a3f078b4c9934e72ce4');">
      <span> Share </span>
    </a>
    <span class="postEdit">
      <span>&middot;</span>
      <a onclick="modalPostEditBox('61e19a3f078b4c9934e72ce4')"> Edit </a>
    </span>
  </div>

  <div class="ext hidden">
    <p>
      15:43 - 14 Jan 2022  &middot; Via
      <a class="via youtube" href="//youtube.com/watch?v=track_A">YouTube</a>
           </p>

    <form class="comments">
    </form>
  </div>
</div>
     </div>
    <!-- end posts -->
  </div>
</div>
<script>
  window.whydCtx = 'hot';
  $(function () {
    $('.post:eq(0) .thumb').append('<div class="medal">1</div>');
    $('.post:eq(1) .thumb').append('<div class="medal">2</div>');
    $('.post:eq(2) .thumb').append('<div class="medal">3</div>');
    /*$(".post .thumb").each(function(i){
			var $post = $(this);
			if (!$post.find(".medal").length)
				$post.append('<div class="medal smallMedal">'+(i+1)+'</div>');
		});
		$(".post:eq(0) .thumb .smallMedal").removeClass('smallMedal').addClass('bigMedal');*/
  });
</script>

  </div>
</div>
<div id="playerContainer">
  <div id="playerContainerSub">
    <div
      id="playerClose"
      onclick="window.whydPlayer.toggleVideo();"
      title="Hide"
    ></div>
    <div id="playBtnOverlay" onclick="window.whydPlayer.playPause();"></div>
  </div>
</div>
<div id="whydPlayer">
  <div id="progressTrack" class="progressTrack">
    <div id="progressBar" class="progressBar"></div>
    <div id="progressCursor" class="progressCursor"></div>
  </div>
  <div class="rightButtons buttons">
    <div
      id="btnShuffle"
      onclick="whydPlayer.toggleShuffle()"
      title="Shuffle mode"
    >
      <div></div>
    </div>
    <div class="volumePanel">
      <div class="volume"></div>
      <div id="volumeTrack" class="progressTrack">
        <div class="progressBar"></div>
        <div class="progressCursor"></div>
      </div>
    </div>
    <div
      id="btnFullscreen"
      onclick="whydPlayer.toggleFullscreen()"
      title="Fullscreen"
    >
      <div></div>
    </div>
  </div>
  <div class="progressPanel buttons">
    <button id="btnPrev" onclick="whydPlayer.prev()"></button>
    <button id="btnPlay" onclick="whydPlayer.playPause()"></button>
    <button id="btnNext" onclick="whydPlayer.next()"></button>
    <button id="btnRepost" onclick="whydPlayer.repost()" title="Add">
      <div></div>
    </button>
    <button id="btnLike" onclick="whydPlayer.like()" title="Like">
      <div></div>
    </button>
    <button id="btnComment" onclick="whydPlayer.comment()" title="Comment">
      <div></div>
    </button>
    <span id="trackTitle">(none)</span>
    <span id="progressTimer"></span>
  </div>
  <div id="trackDragPos"></div>
</div>




<script>
var DEEZER_APP_ID = 190482;
var DEEZER_CHANNEL_URL = window.location.href.substr(0, window.location.href.indexOf("/", 10)) + "/html/channel.html";
var YOUTUBE_API_KEY = "AIzaSyBzqb519R--gKZ9cXgbqE5bMM85yNTXJfo";
var JAMENDO_CLIENT_ID = "2c9a11b9";
</script>
    <script src="/js/jquery.avgrund.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/jquery.tipsy.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/quickSearch.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/jquery.iframe-post-form.min.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/jquery.placeholder.min.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/underscore-min.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/jquery.elastic.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/jquery.mentionsInput.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/ui.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/whyd.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/playem-all.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/playem-youtube-iframe-patch.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/whydPlayer.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/dndUpload.js" type="text/javascript" charset="utf-8"></script>
    <script src="/js/WhydImgUpload.js" type="text/javascript" charset="utf-8"></script>
  </body>
</html>"
`;
